### ISSUE TITLE
Make Maven Artifact cache configurable

### ISSUE THREAD

**Issue by fabapp2** [spring-projects-experimental/spring-boot-migrator#111]
### What needs to be done
Make Maven artifact cache configurable by setting an application property.

#### Read from Maven Repository but download new dependencies to separate Maven cache (current)
````
new MavenArtifactDownloader(
                        ReadOnlyLocalMavenArtifactCache.mavenLocal().orElse(
                            new LocalMavenArtifactCache(Paths.get(System.getProperty("user.home"), ".rewrite", "cache", "artifacts")
````

#### Use local Maven repository for reading and writing
````
new MavenArtifactDownloader(new LocalMavenArtifactCache(Paths.get(System.getProperty("user.home"), ".m2", "repository")), ...
````

### Why it needs to be done
It can be useful (and easier to understand) if all dependencies are stored and retrieved from the local Maven repository but then SBM potentially overwrites dependencies in the local Maven repository.
Allow a user to choose.

### Additional Information
See `DependencyHelper.downloadArtifacts(...)` for artifact configuration and `ApplicationProperties` for properties

**Comment by ravig-kant** [comment]
@fabapp2 I would like to take this up.

**Comment by fabapp2** [comment]
> @fabapp2 I would like to take this up.
Hi @ravig-kant,
that would be awesome, thanks for your interest!

Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring) beforehand, if you haven't already done so.

I just checked again and realized it will require some refactoring to make `DependencyHelper` a spring-managed bean as it's currently created using `new`. If you don't mind, it would be helpful as a first step to just try to implement the switch inside `DependencyHelper` depending on the property, e.g. `sbm.onlyUseLocalMavenRepository` (or a better name) and then I/we see how the `DependencyHelper` can be made a spring-managed bean in a second step?

The relevant branch (e.g. with related TODO in `DependencyHelper`) is `upgrade-rewrite-7.18.2` which I am working on atm. Please create a new branch from this branch in your fork so we can use this for discussion.

Please also have a look at `OpenRewriteMavenBuildFileTest.testResolvedDependenciesWithPomTypeDependency()`
where this problem surfaced to me. In this test, I hacked something to deal with the current implementation. This could be changed afterward to use the new "Use local Maven repository for reading and writing" configuration.
Basically by adding matching application properties like this
````
TestProjectContext.buildProjectContext()
                .withApplicationProperties(...)
                ...
````
Which would allow to remove the hack in `OpenRewriteMavenBuildFileTest.testResolvedDependenciesWithPomTypeDependency`
````
if(dep.contains(".rewrite/cache/artifacts/")) {
                        return dep.substring(dep.lastIndexOf(".rewrite/cache/artifacts/") + ".rewrite/cache/artifacts/".length());
                    } else {
````

If everything works and we found a way how to make `DependencyHelper` a spring-bean, a section in `applications/spring-shell/src/main/resources/banner.txt` describing the functionality and the current property setting would be nice.

If you have questions or need help, please don't hesitate to ask.

**Comment by ravig-kant** [comment]
Thanks for the detailed explanation. Actually, I started working on making DependencyHelper a spring bean before your email. I should have replied earlier, but I got caught up in some personal work.

The approach I took was to create an inner class Builder in ClasspathRegistry and make it a spring Bean.  DependencyHelper will be autowired in the Builder. The Builder then uses the injected DependencyHelper to initDependency in the ClasspathRegistry. 
ClasspathRegistry instance is created by OpenRewriteMavenBuildFile which is instantiated from spring beans. So I thought of injecting ClasspathRegistry.Builder to OpenRewriteMavenBuildFile. 
Please let me know whether this approach is correct. Otherwise, I will use a simple property-based approach for now. I am almost halfway through with changes. I can quickly add it to my fork so that you can review the code. It would be much easier to understand the approach by looking at the code.

**Comment by fabapp2** [comment]
Pushing it into your fork would be helpful. I think I vaguely understood but looking at the code would ease our conversation.
If you like you could also create a draft P R and we discuss it there. Both work for me. Please be aware that I will be ooo from  11th for 7 days. Will check tomorrow, otherwise when I'm back

**Comment by ravig-kant** [comment]
Hi @fabapp2 

I pulled the latest branch, and it seems that my original idea of introducing the Builder requires another round of thinking with new changes. For example, the earlier version didn't have JavaParsers and Current and Initial dependencies concepts.
In a week, I will either come up with a proposal on how to make Dependency Helper a spring bean or have a PR with a simple property. I want to use this opportunity to learn about the code to contribute more in the future.

**Comment by fabapp2** [comment]
Hi @ravig-kant 
I am happy about your interest and any contribution is appreciated!

While working on the upgrade to OpenRewrite 7.23.0 I realized some changes that will probably affect your work, I'm afraid might have been tagging this as 'good-first-issue' a bit too fast.
Nevertheless, I would be glad to have you contribute but unfortunately there are a few issues that need to be solved first to not waste your precious time on this. Maybe you want to pick something else for now, or maybe just wait until I'm back from PTO next week? 

This needs to be sorted out:

- Quite a few things changed related to Maven and dependency resolution since 7.17.3 which is the Rewrite version currently on `main`
- OpenRewrite's `AddDependency` recipe can now download added dependencies
- Being able to replace SBM's `AddDependency` recipe with OR's would make the `DependencyHelper` used in `OpenRewriteMavenBuildFile.addDependenc` obsolete as it's used to download dependencies added to the pom.xml
- Then the `MavenArtifactDownloader` wouldn't be used in `DependencyHelper` anymore
- **But** OR's `AddDependency` requires a type in use `onlyIfUsing` to be effective. In SBM the type must be added after adding the dependency
- To switch between the two modes (rewrite cache or Maven cache) the `MavenArtifactCache` used in `MavenArtifactDownloader` needs to be reconfigured/changed by an application property
- `MavenArtifactDownloader` is currently only used in `MavenProjectParser` when initially parsing the project
- **Q:** Is the `MavenArtifactDownloader` used to download dependencies?

**Comment by fabapp2** [comment]
> Hi @fabapp2
> 
> I pulled the latest branch, and it seems that my original idea of introducing the Builder requires another round of thinking with new changes. For example, the earlier version didn't have JavaParsers and Current and Initial dependencies concepts. In a week, I will either come up with a proposal on how to make Dependency Helper a spring bean or have a PR with a simple property. I want to use this opportunity to learn about the code to contribute more in the future.

If you're keen to work on this, having a simple property in `application-core.properties` mapped to `ApplicationProperties` would be a start to getting your PR merged.
Everything else needs more thought from my side, unfortunately :)

**Comment by ravig-kant** [comment]
Hi @fabapp2 

I have raised a PR for the property. In the meantime, I am going through the code to understand it better and be ready for bigger contributions.

**Comment by fabapp2** [comment]
closed with #912
